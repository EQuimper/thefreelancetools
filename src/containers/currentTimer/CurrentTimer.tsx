import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Button, Combobox, Text, TextInput } from 'evergreen-ui';
import { observer } from 'mobx-react';
import * as R from 'ramda';
import * as React from 'react';

import { store, Task } from '@freelance-tool/models';
import { IconEnum } from '@freelance-tool/types';

interface OptionItem {
  label: string;
  value: string;
}

interface P {}

interface S {
  selectedProject: OptionItem | null;
  taskName: string;
}

@observer
class CurrentTimer extends React.Component<P, S> {
  state = {
    selectedProject: null,
    taskName: '',
  };

  _handleStart = () => {
    const getId = R.pathOr('', ['selectedProject', 'value']);

    const project = store.projects.getProjectById(getId(this.state));

    if (project) {
      const task = Task.create({
        name: this.state.taskName,
        elapsedTime: {
          hours: 0,
          minutes: 0,
          seconds: 0,
          totalSeconds: 0,
        },
      });

      project.addTask(task);
      store.currentTimer.start(project, task);
    }
  }

  _handleProjectChange = (value: OptionItem) => {
    this.setState({
      selectedProject: value,
    });
  }

  render() {
    const { currentTimer, projects } = store;

    return (
      <div>
        <Combobox
          placeholder="Project"
          autocompleteProps={{
            title: 'Project Name',
            itemsFilter: (items: OptionItem[], input: string) =>
              items.filter(el => el.label.includes(input)),
          }}
          openOnFocus
          items={projects.getProjectsOptions}
          onChange={this._handleProjectChange}
          itemToString={(i: OptionItem) => (i ? i.label : '')}
          selectedItem={this.state.selectedProject}
        />

        <br />
        <br />
        <TextInput
          placeholder="Task name"
          onChange={(e: InputEvent) =>
            this.setState({ taskName: e.target.value })
          }
          value={this.state.taskName}
        />
        <br />
        <br />

        <Text>ElapseTime {currentTimer.getCurrentElapseTime}</Text>
        {currentTimer.isRunning ? (
          <Button
            appearance="ghost"
            marginLeft={12}
            onClick={currentTimer.stop}
          >
            <FontAwesomeIcon size="2x" icon={IconEnum.PAUSE} />
          </Button>
        ) : (
          <Button
            appearance="ghost"
            marginLeft={12}
            onClick={this._handleStart}
          >
            <FontAwesomeIcon size="2x" icon={IconEnum.PLAY} />
          </Button>
        )}

        <Button
          marginLeft={12}
          appearance="ghost"
          onClick={currentTimer.finish}
        >
          Finish
        </Button>
        <Button
          marginLeft={12}
          appearance="ghostBlue"
          onClick={currentTimer.reset}
        >
          Reset
        </Button>
      </div>
    );
  }
}

export default CurrentTimer;
