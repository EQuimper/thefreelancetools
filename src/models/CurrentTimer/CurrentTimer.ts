import { types } from 'mobx-state-tree';

export const CurrentTimer = types
  .model('CurrentTimer', {
    elapseTime: types.model({
      hours: 0,
      minutes: 0,
      seconds: 0,
    }),
    isRunning: false,
    intervalId: types.maybe(types.number),
  })
  .actions(self => ({
    updateElapseTime() {
      self.elapseTime.seconds += 1;

      if (self.elapseTime.seconds === 60) {
        self.elapseTime.minutes += 1;
        self.elapseTime.seconds = 0;

        if (self.elapseTime.minutes === 60) {
          self.elapseTime.hours += 1;
          self.elapseTime.minutes = 0;
        }
      }
    },
  }))
  .actions(self => ({
    start() {
      self.intervalId = Number(setInterval(self.updateElapseTime, 1000));
      self.isRunning = true;
    },
  }))
  .actions(self => ({
    stop() {
      if (self.intervalId) {
        clearInterval(self.intervalId);
      }
      self.intervalId = null;
      self.isRunning = false;
    },
  }))
  .actions(self => ({
    reset() {
      self.elapseTime = {
        hours: 0,
        minutes: 0,
        seconds: 0,
      };
    },
  }))
  .views(self => ({
    get getCurrentElapseTime() {
      return self.elapseTime;
    },
  }));
