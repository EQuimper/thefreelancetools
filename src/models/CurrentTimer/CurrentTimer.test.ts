import { onSnapshot } from 'mobx-state-tree';

import { CurrentTimer } from '../CurrentTimer';

jest.useFakeTimers();

describe('CurrentTimer model', () => {
  it('should create a instance of a model', () => {
    const currentTimer = CurrentTimer.create({
      elapseTime: {
        hours: 0,
        minutes: 0,
        seconds: 0,
      },
    });

    expect(currentTimer.isRunning).toBe(false);
    expect(currentTimer.elapseTime).toEqual({
      hours: 0,
      minutes: 0,
      seconds: 0,
    });
    expect(currentTimer.intervalId).toBe(null);
  });

  it('should be able to start', () => {
    const currentTimer = CurrentTimer.create({
      elapseTime: {
        hours: 0,
        minutes: 0,
        seconds: 0,
      },
    });

    currentTimer.start();

    expect(currentTimer.isRunning).toBe(true);
    expect(currentTimer.intervalId).toBe(1);
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);
  });

  it('should be able to stop', () => {
    const currentTimer = CurrentTimer.create({
      elapseTime: {
        hours: 0,
        minutes: 0,
        seconds: 0,
      },
    });

    const states: any[] = [];

    onSnapshot(currentTimer, snapshot => {
      states.push(snapshot);
    });

    currentTimer.start();

    expect(setInterval).toHaveBeenCalledTimes(2);

    jest.runOnlyPendingTimers();

    expect(currentTimer.elapseTime).not.toBe(0);

    currentTimer.stop();

    expect(currentTimer.isRunning).toBe(false);
    expect(currentTimer.intervalId).toBe(null);

    expect(states).toMatchSnapshot();
  });

  it('should be able to reset', () => {
    const currentTimer = CurrentTimer.create({
      elapseTime: {
        hours: 0,
        minutes: 0,
        seconds: 0,
      },
    });

    const states: any[] = [];

    onSnapshot(currentTimer, snapshot => {
      states.push(snapshot);
    });

    currentTimer.start();

    expect(setInterval).toHaveBeenCalledTimes(3);

    jest.runOnlyPendingTimers();

    expect(currentTimer.elapseTime).not.toBe(0);

    currentTimer.stop();

    expect(currentTimer.isRunning).toBe(false);
    expect(currentTimer.intervalId).toBe(null);

    currentTimer.start();

    jest.runOnlyPendingTimers();

    currentTimer.stop();

    currentTimer.reset();

    expect(states).toMatchSnapshot();
  });
});
