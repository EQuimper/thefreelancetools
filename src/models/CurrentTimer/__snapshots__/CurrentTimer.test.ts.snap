// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentTimer model should be able to reset 1`] = `
Array [
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 0,
      "totalSeconds": 0,
    },
    "intervalId": 14,
    "isRunning": true,
    "project": "123",
    "task": "999",
  },
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 1,
      "totalSeconds": 1,
    },
    "intervalId": 14,
    "isRunning": true,
    "project": "123",
    "task": "999",
  },
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 1,
      "totalSeconds": 1,
    },
    "intervalId": null,
    "isRunning": false,
    "project": "123",
    "task": "999",
  },
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 1,
      "totalSeconds": 1,
    },
    "intervalId": 16,
    "isRunning": true,
    "project": "123",
    "task": "999",
  },
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 2,
      "totalSeconds": 2,
    },
    "intervalId": 16,
    "isRunning": true,
    "project": "123",
    "task": "999",
  },
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 2,
      "totalSeconds": 2,
    },
    "intervalId": null,
    "isRunning": false,
    "project": "123",
    "task": "999",
  },
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 0,
      "totalSeconds": 0,
    },
    "intervalId": null,
    "isRunning": false,
    "project": "123",
    "task": "999",
  },
]
`;

exports[`CurrentTimer model should be able to stop 1`] = `
Array [
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 0,
      "totalSeconds": 0,
    },
    "intervalId": 8,
    "isRunning": true,
    "project": "123",
    "task": "999",
  },
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 1,
      "totalSeconds": 1,
    },
    "intervalId": 8,
    "isRunning": true,
    "project": "123",
    "task": "999",
  },
  Object {
    "elapseTime": Object {
      "hours": 0,
      "minutes": 0,
      "seconds": 1,
      "totalSeconds": 1,
    },
    "intervalId": null,
    "isRunning": false,
    "project": "123",
    "task": "999",
  },
]
`;
